server/data/restClient.ts

```diff
@@ -81,62 +90,68 @@ export default class RestClient {
+    return this.processRequest({ path, headers, responseType, data, raw, method: 'post' })
+  }
 
-      return raw ? result : result.body
-    } catch (error) {
-      const sanitisedError = sanitiseError(error)
-      logger.warn({ ...sanitisedError }, `Error calling ${this.name}, path: '${path}', verb: 'POST'`)
-      throw sanitisedError
-    }
+  async put({
+    path = null,
+    headers = {},
+    responseType = '',
+    data = {},
+    raw = false,
+  }: PutRequest = {}): Promise<unknown> {
+    return this.processRequest({ path, headers, responseType, data, raw, method: 'put' })
   }
 
-  async stream({ path = null, headers = {} }: StreamRequest = {}): Promise<unknown> {
-    logger.info(`Get using user credentials: calling ${this.name}: ${path}`)
-    return new Promise((resolve, reject) => {
-      superagent
-        .get(`${this.apiUrl()}${path}`)
-        .agent(this.agent)
-        .auth(this.token, { type: 'bearer' })
-        .use(restClientMetricsMiddleware)
-        .retry(2, (err, res) => {
-          if (err) logger.info(`Retry handler found API error with ${err.code} ${err.message}`)
-          return undefined // retry handler only for logging retries, not to influence retry logic
-        })
-        .timeout(this.timeoutConfig())
-        .set(headers)
-        .end((error, response) => {
-          if (error) {
-            logger.warn(sanitiseError(error), `Error calling ${this.name}`)
-            reject(error)
-          } else if (response) {
-            const s = new Readable()
-            // eslint-disable-next-line no-underscore-dangle,@typescript-eslint/no-empty-function
-            s._read = () => {}
-            s.push(response.body)
-            s.push(null)
-            resolve(s)
-          }
-        })
-    })
+  async delete({
+    path = null,
+    headers = {},
+    responseType = '',
+    data = {},
+    raw = false,
+  }: DeleteRequest = {}): Promise<unknown> {
+    return this.processRequest({ path, headers, responseType, data, raw, method: 'delete' })
+  }
+
+  private async processRequest({
+    path = null,
+    headers = {},
+    responseType = '',
+    query = undefined,
+    data = undefined,
+    raw = false,
+    method = undefined as 'get' | 'post' | 'put' | 'delete',
+  } = {}): Promise<unknown> {
+    const request = superagent[method](`${this.apiUrl()}${path}`)
+    request
+      .agent(this.agent)
+      .retry(2, (err: ResponseError): void => {
+        if (err) logger.info(`Retry handler found API error with ${err.status} ${err.message}`)
+        return undefined // retry handler only for logging retries, not to influence retry logic
+      })
+      .set(headers)
+      .responseType(responseType)
+      .timeout(this.timeoutConfig())
+
+    if (data) {
+      request.send(data)
+    }
+    if (query) {
+      request.query(query)
+    }
+
+    request.auth(this.token, { type: 'bearer' })
+    if (this.sourceIp) {
+      request.set('x-slm-client-ip', this.sourceIp)
```

Comment:
I think this file is _probably_ OK but now that CJSM stuff is gone it seems like it's only adding `x-slm-client-ip` that is different to the template app?

So worth considering if this could go back to something more like the default implementation you started with in this repo. Because you've lost some recent improvements to the template's rest client - e.g. no retries for `POST / PUT`. (There's a very recent change to the template's rest client - last week - but that might be a bigger change.)

Reply:
The other thing it does differently is on get requests to PrisonService it does not pass a token but passes a token for other get requests. We can have a call tomorrow to discuss.

Reply:
I have updated the existing restClient to match the new template with subtle changes. Can you please have a look. We can have a call to discuss as well.